// @tanstack/query-db-collection@0.1.3/dist/cjs/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const query = require("./query.cjs");
const errors = require("./errors.cjs");
exports.queryCollectionOptions = query.queryCollectionOptions;
exports.DeleteOperationItemNotFoundError = errors.DeleteOperationItemNotFoundError;
exports.DuplicateKeyInBatchError = errors.DuplicateKeyInBatchError;
exports.GetKeyRequiredError = errors.GetKeyRequiredError;
exports.InvalidItemStructureError = errors.InvalidItemStructureError;
exports.InvalidSyncOperationError = errors.InvalidSyncOperationError;
exports.ItemNotFoundError = errors.ItemNotFoundError;
exports.MissingKeyFieldError = errors.MissingKeyFieldError;
exports.QueryClientRequiredError = errors.QueryClientRequiredError;
exports.QueryCollectionError = errors.QueryCollectionError;
exports.QueryFnRequiredError = errors.QueryFnRequiredError;
exports.QueryKeyRequiredError = errors.QueryKeyRequiredError;
exports.SyncNotInitializedError = errors.SyncNotInitializedError;
exports.UnknownOperationTypeError = errors.UnknownOperationTypeError;
exports.UpdateOperationItemNotFoundError = errors.UpdateOperationItemNotFoundError;
